torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -975,6 +1023,10 @@ def __init__(
         self._materialize_module(module, param_init_fn, device_from_device_id)
         self._move_module_to_device(module, ignored_params, device_from_device_id)
         self.compute_device = self._get_compute_device(module, ignored_params, device_from_device_id)
+        self._max_inflight_all_gather_size = (
+            torch.cuda.get_device_properties(self.compute_device).total_memory
+            * 0.005  # empirically chosen, e.g. 200 MB limit for 40 GB GPU
```

Comment:
assume we should choose larger limit to avoid frequently slow down the CPU thread in default, otherwise slow CPU may result in larger cuda kernel scheduling overhead and slow down training?

Reply:
I have been experimenting with this (but much more to come). I originally chose `0.05`, i.e. 2 GB limit, but it is basically not effective then.

The observation is that when people scale models, they add more layers, not increase the size of each layer. (For example, think of transformer layers and the transformer auto wrap policy.) This means that the size of each `FlatParameter` is actually pretty fixed when scaling the model, at least for recursive wrapping. For DeepViT, each padded unsharded flattened parameter is 16 MB. If we set a limit of 2 GB, it will take too many all-gathers to get hit, and the limit is not effective. The CPU thread will have already ran ahead.

We can keep brainstorming how to decide this limit. For this particular approach of using a scaling factor against the total GPU memory, we may be able to use the [elbow method](https://en.wikipedia.org/wiki/Elbow_method_(clustering)).